server/memcached/src/main/java/org/infinispan/server/memcached/text/TextOpDecoder.java

```diff
@@ -64,35 +68,48 @@ protected TextOpDecoder(MemcachedServer server, Subject subject) {
       super(server, subject);
    }
 
-   protected void get(TextHeader header, List<byte[]> keys, boolean withVersions) {
+   protected MemcachedResponse get(TextHeader header, List<byte[]> keys, boolean withVersions) {
       int numberOfKeys = keys.size();
       if (numberOfKeys > 1) {
          CacheEntry<byte[], byte[]>[] entries = new CacheEntry[numberOfKeys];
-         CompletionStage<Void> lastStage = doGetMultipleKeys(keys, entries, 0);
-         for (int i = 1; i < numberOfKeys; ++i) {
-            final int idx = i;
-            lastStage = lastStage.thenCompose(unused -> doGetMultipleKeys(keys, entries, idx));
+         AggregateCompletionStage<CacheEntry<byte[], byte[]>[]> acs = CompletionStages.aggregateCompletionStage(entries);
+         for (int i = 0; i < numberOfKeys; ++i) {
+            acs.dependsOn(doGetMultipleKeys(keys, entries, i));
          }
-         send(header, lastStage.thenApply(unused -> createMultiGetResponse(entries)));
-      } else {
-         byte[] key = keys.get(0);
-         send(header, cache.getCacheEntryAsync(key).thenApply(entry -> createGetResponse(key, entry, withVersions)));
+
+         CompletionStage<CacheEntry<byte[], byte[]>[]> cs = acs.freeze();
+         if (CompletionStages.isCompletedSuccessfully(cs)) {
+            return send(header, CompletableFuture.completedFuture(createMultiGetResponse(entries)));
+         }
+
+         return send(header, cs.thenApply(unused -> createMultiGetResponse(entries)));
```

Comment:
```suggestion
         return send(header, cs.thenApply(innerEntries -> createMultiGetResponse(innerEntries)));
```

I think this was the original intention of storing the entries in the AggregateCompletionstage.

Also I just noticed that the writing of entries is not synchronized, so I think this is technically an issue that the invoking thread here may not 100% see the value.

Reply:
Updated. I am wrapping everything in a synchronized list.